velox/functions/prestosql/InPredicate.cpp

```diff
@@ -86,6 +86,120 @@ std::pair<std::unique_ptr<common::Filter>, bool> createBigintValuesFilter(
+  values.reserve(size);
+  bool nullAllowed = false;
+
+  for (auto i = offset; i < offset + size; i++) {
+    if (elementsVector->isNullAt(i)) {
+      nullAllowed = true;
+    } else {
+      values.emplace_back(elementsVector->valueAt(i).unscaledValue());
+    }
+  }
+
+  return std::optional<std::pair<std::vector<int64_t>, bool>>(
+      {std::move(values), nullAllowed});
+}
+
+std::pair<std::unique_ptr<common::Filter>, bool> createShortDecimalValuesFilter(
+    const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  auto valuesPair = toUnscaledShortDecimalValues(inputArgs);
+  if (!valuesPair.has_value()) {
+    return {nullptr, false};
+  }
+
+  const auto& values = valuesPair.value().first;
+  bool nullAllowed = valuesPair.value().second;
+
+  if (values.empty() && nullAllowed) {
+    return {nullptr, true};
+  }
+  VELOX_USER_CHECK(
+      !values.empty(),
+      "IN predicate expects at least one non-null value in the in-list");
+
+  return {common::createBigintValues(values, nullAllowed), false};
+}
+
+std::optional<std::pair<std::vector<int128_t>, bool>>
+toUnscaledLongDecimalValues(
+    const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  const auto& constantValue = inputArgs[1].constantValue;
+  if (!constantValue) {
+    return std::nullopt;
+  }
+
+  auto constantInput =
+      std::dynamic_pointer_cast<ConstantVector<ComplexType>>(constantValue);
+  if (constantInput->isNullAt(0)) {
+    return std::optional<std::pair<std::vector<int128_t>, bool>>({{}, true});
+  }
+  auto arrayVector = dynamic_cast<const ArrayVector*>(
+      constantInput->valueVector()->wrappedVector());
+  auto elementsVector =
+      arrayVector->elements()->as<SimpleVector<UnscaledShortDecimal>>();
+  auto offset = arrayVector->offsetAt(constantInput->index());
+  auto size = arrayVector->sizeAt(constantInput->index());
+  VELOX_USER_CHECK_GT(size, 0, "IN list must not be empty");
+
+  std::vector<int128_t> values;
+  values.reserve(size);
+  bool nullAllowed = false;
+
+  for (auto i = offset; i < offset + size; i++) {
+    if (elementsVector->isNullAt(i)) {
+      nullAllowed = true;
+    } else {
+      values.emplace_back(elementsVector->valueAt(i).unscaledValue());
+    }
+  }
+
+  return std::optional<std::pair<std::vector<int128_t>, bool>>(
+      {std::move(values), nullAllowed});
+}
+
+std::pair<std::unique_ptr<common::Filter>, bool> createLongDecimalValuesFilter(
+    const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  auto valuesPair = toUnscaledLongDecimalValues(inputArgs);
+  if (!valuesPair.has_value()) {
+    return {nullptr, false};
+  }
+
+  const auto& values = valuesPair.value().first;
+  bool nullAllowed = valuesPair.value().second;
+
+  if (values.empty() && nullAllowed) {
+    return {nullptr, true};
+  }
+  VELOX_USER_CHECK(
+      !values.empty(),
+      "IN predicate expects at least one non-null value in the in-list");
+
+  return {common::createHugeintValues(values, nullAllowed), false};
```

Comment:
Same as above,  this method seems exactly same as the other. Try to combine and minimize code.